File: apps/docker-server/cmd/renewlet/server_i18n.go
package main

import (
	"embed"
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"golang.org/x/text/language"
)

// Go catalog 与 Worker catalog 由 shared JSON 同源生成；前端错误展示只共享 code，不共享服务端翻译文本。
//
//go:embed i18n/active.*.json
var serverI18nFS embed.FS

var (
	serverI18nCatalogs = mustLoadServerI18nCatalogs()
	serverI18nLocales  = append([]appLocale(nil), supportedAppLocales...)
	serverI18nTags     = serverI18nLanguageTags(serverI18nLocales)
	serverI18nMatcher  = language.NewMatcher(serverI18nTags)
	acceptLanguageQRe  = regexp.MustCompile(`^(?:0(?:\.[0-9]{0,3})?|1(?:\.0{0,3})?)$`)
)

func mustLoadServerI18nCatalogs() map[appLocale]map[string]string {
	catalogs := map[appLocale]map[string]string{}
	for _, locale := range supportedAppLocales {
		data, err := serverI18nFS.ReadFile("i18n/active." + string(locale) + ".json")
		if err != nil {
			panic(err)
		}
		var catalog map[string]string
		if err := json.Unmarshal(data, &catalog); err != nil {
			panic(err)
		}
		catalogs[locale] = catalog
	}
	if _, ok := catalogs[defaultAppLocale]; !ok {
		panic(fmt.Sprintf("missing default server i18n catalog %s", defaultAppLocale))
	}
	return catalogs
}

func serverI18nLanguageTags(locales []appLocale) []language.Tag {
	tags := make([]language.Tag, 0, len(locales))
	for _, locale := range locales {
		tags = append(tags, language.MustParse(string(locale)))
	}
	return tags
}

// matcher 只接受合法语言标签并按受支持语言收敛；这里的结果必须与 Worker matchServerLocale 保持同构。
func matchAppLocale(value string) (appLocale, bool) {
	value = strings.TrimSpace(strings.ReplaceAll(value, "_", "-"))
	if value == "" {
		return defaultAppLocale, false
	}
	tag, err := language.Parse(value)
	if err != nil {
		return defaultAppLocale, false
	}
	_, index, confidence := serverI18nMatcher.Match(tag)
	if confidence == language.No {
		return defaultAppLocale, false
	}
	return serverI18nLocales[index], true
}

func normalizeAppLocale(value string) appLocale {
	if locale, ok := matchAppLocale(value); ok {
		return locale
	}
	return defaultAppLocale
}

func isSupportedAppLocale(value string) bool {
	for _, locale := range supportedAppLocales {
		if value == string(locale) {
			return true
		}
	}
	return false
}

// Accept-Language 只服务无显式偏好请求；按 q 权重和原始顺序选择，忽略非法项，无匹配或通配符回退英文。
// 解析规则必须与 Worker requestLocale 的夹具保持一致，避免 Docker/Cloudflare 返回不同语言。
func matchAcceptLanguage(header string) appLocale {
	type preference struct {
		tag   string
		q     float64
		index int
	}
	preferences := []preference{}
	for index, rawPart := range strings.Split(header, ",") {
		parts := strings.Split(rawPart, ";")
		tag := strings.TrimSpace(parts[0])
		if tag == "" {
			continue
		}
		quality := 1.0
		valid := true
		for _, rawParameter := range parts[1:] {
			parameter := strings.TrimSpace(rawParameter)
			if len(parameter) < 2 || !strings.EqualFold(parameter[:2], "q=") {
				continue
			}
			value := strings.TrimSpace(parameter[2:])
			if !acceptLanguageQRe.MatchString(value) {
				valid = false
				break
			}
			parsed, err := strconv.ParseFloat(value, 64)
			if err != nil || parsed <= 0 {
				valid = false
				break
			}
			quality = parsed
			break
		}
		if valid {
			preferences = append(preferences, preference{tag: tag, q: quality, index: index})
		}
	}
	sort.SliceStable(preferences, func(i, j int) bool {
		if preferences[i].q == preferences[j].q {
			return preferences[i].index < preferences[j].index
		}
		return preferences[i].q > preferences[j].q
	})
	for _, candidate := range preferences {
		if strings.TrimSpace(candidate.tag) == "*" {
			return defaultAppLocale
		}
		if matched, ok := matchAppLocale(candidate.tag); ok {
			return matched
		}
	}
	return defaultAppLocale
}

// 缺少目标语言或 key 时统一回退英文 catalog；最终返回 key 让缺失构件在测试和日志中可见。
func serverText(locale appLocale, key string) string {
	if catalog, ok := serverI18nCatalogs[locale]; ok {
		if message, ok := catalog[key]; ok {
			return message
		}
	}
	if message, ok := serverI18nCatalogs[defaultAppLocale][key]; ok {
		return message
	}
	return key
}

func serverFormat(locale appLocale, key string, params map[string]interface{}) string {
	return serverFormatMessage(serverText(locale, key), params)
}

func serverFormatMessage(message string, params map[string]interface{}) string {
	for name, value := range params {
		message = strings.ReplaceAll(message, "{"+name+"}", fmt.Sprint(value))
	}
	return message
}

func localizedDisabledBanReason(locale appLocale) string {
	return serverText(locale, "auth.accountDisabledByAdmin")
}

File: apps/worker/src/server-i18n.ts
import { resolveLocalePreference, type LocalePreference } from "@renewlet/shared/i18n-config";
import {
  DEFAULT_SERVER_I18N_LOCALE,
  SERVER_I18N_CATALOGS,
  SERVER_I18N_LOCALES,
  type ServerI18nCatalog,
  type ServerI18nKey,
  type ServerI18nLocale,
} from "./server-i18n-catalog";

export type AppLocale = ServerI18nLocale;
export { DEFAULT_SERVER_I18N_LOCALE };

// Worker 服务端文案使用生成 catalog，不复用浏览器 Lingui runtime；前端只消费稳定错误 code。

const localeLookup = new Map<string, AppLocale>();
const acceptLanguageQPattern = /^(?:0(?:\.[0-9]{0,3})?|1(?:\.0{0,3})?)$/;
for (const locale of SERVER_I18N_LOCALES) {
  const normalized = normalizeLocaleTag(locale);
  localeLookup.set(normalized, locale);
  const language = normalized.split("-")[0] ?? normalized;
  if (!localeLookup.has(language)) localeLookup.set(language, locale);
}

function normalizeLocaleTag(value: string): string {
  const normalized = value.trim().replaceAll("_", "-");
  if (!normalized) return "";
  try {
    return (Intl.getCanonicalLocales(normalized)[0] ?? "").toLowerCase();
  } catch {
    return "";
  }
}

function acceptLanguageQuality(value: string | undefined): number {
  if (value === undefined) return 1;
  const normalized = value.trim();
  return acceptLanguageQPattern.test(normalized) ? Number(normalized) : Number.NaN;
}

// 合法完整标签先精确匹配，再按基础语言收敛；结果必须与 Go matchAppLocale 保持同构。
function matchServerLocale(value: string | null | undefined): AppLocale | null {
  const normalized = normalizeLocaleTag(value ?? "");
  if (!normalized) return null;
  const direct = localeLookup.get(normalized);
  if (direct) return direct;
  const language = normalized.split("-")[0] ?? normalized;
  return localeLookup.get(language) ?? null;
}

export function normalizeServerLocale(value: string | null | undefined): AppLocale {
  return matchServerLocale(value) ?? DEFAULT_SERVER_I18N_LOCALE;
}

/** 请求语言与账号内容语言互不覆盖；auto 在 Cron、Feed、Bot 等无设备上下文中稳定回退英文。 */
export function accountContentLocale(preference: LocalePreference): AppLocale {
  return resolveLocalePreference(preference);
}

/** requestLocale 优先读取前端随用户设置发送的显式 locale header。 */
export function requestLocale(request: Request): AppLocale {
  const explicit = request.headers.get("x-renewlet-locale");
  if (explicit?.trim()) {
    // 显式 header 来自用户设置，比浏览器语言更可信；非法值只回默认语言，不再被 Accept-Language 反向覆盖。
    const value = explicit.trim();
    return (SERVER_I18N_LOCALES as readonly string[]).includes(value)
      ? value as AppLocale
      : DEFAULT_SERVER_I18N_LOCALE;
  }
  // Accept-Language 只是无显式设置时的兜底；q 权重、非法项和通配符规则必须与 Go matcher 同构。
  const accepted = (request.headers.get("accept-language") ?? "")
    .split(",")
    .map((part, index) => {
      const [tag = "", ...params] = part.trim().split(";");
      const qValue = params.map((item) => item.trim()).find((item) => item.slice(0, 2).toLowerCase() === "q=")?.slice(2);
      return { tag: tag.trim(), q: acceptLanguageQuality(qValue), index };
    })
    .filter((item) => item.tag && Number.isFinite(item.q) && item.q > 0 && item.q <= 1)
    .sort((a, b) => b.q - a.q || a.index - b.index);
  for (const { tag } of accepted) {
    if (tag === "*") return DEFAULT_SERVER_I18N_LOCALE;
    const matched = matchServerLocale(tag);
    if (matched) return matched;
  }
  return DEFAULT_SERVER_I18N_LOCALE;
}

/** serverText 返回服务端 catalog 文案；缺 key 时回 key，便于测试发现 catalog 漂移。 */
export function serverText(locale: AppLocale, key: ServerI18nKey): string {
  const catalogs = SERVER_I18N_CATALOGS as Record<AppLocale, ServerI18nCatalog>;
  return catalogs[locale]?.[key] ?? catalogs[DEFAULT_SERVER_I18N_LOCALE][key] ?? key;
}

/** serverFormat 只做服务端错误/通知所需的命名占位替换，不引入前端 i18n runtime。 */
export function serverFormat(locale: AppLocale, key: ServerI18nKey, params: Record<string, string | number>): string {
  return serverText(locale, key).replace(/\{([A-Za-z_][A-Za-z0-9_]*)\}/g, (match, name: string) => {
    const value = params[name];
    return value === undefined ? match : String(value);
  });
}
